Sampler rework: refactor CWMH sampler

amal-ghamdi opened the pull request:
closes #371 

Things to consider for the review:
- The code is mainly copied from the original `CWMH`
- Modification to remove all the outer sampling loops and just keep the step and tune parts
- I needed to introduce additional scale parameter to keep track of old scale (to enable matching the old and new sampler)
- I modified the class documentation a bit, including stating that scale can be an array in this CWMH case. And updated the example code in the class documentation
-  I modified the tests method `assert_true_if_sampling_is_equivalent` and `assert_true_if_warmup_is_equivalent` to accept additional inputs: `old_idx` and `new_idx`. Which I use for example to remove first sample in the new sampler which is the initial point, that is not included in the samples of the old sampler. I use it also to remove the last sample in the old sampler because it is repeated (same as second to last sample)

Question:
- [x] I am not sure about the dim property? it seems like it was remover from the `NewSampler`. Should we still support it?(I removed dim from CWMH as discussed)

nabriis commented on cuqi/experimental/mcmc/_cwmh.py:
```diff
@@ -0,0 +1,207 @@
+    initial_point : ndarray
+        Initial parameters. *Optional*
+
+    dim : int
```
I would suggest we remove the dim property. Instead user can define UserDefinedDistribution. We can discuss at next meeting.

nabriis replied:
After our discussion, we agreed to remove the "dim" property yes.

nabriis commented on cuqi/experimental/mcmc/_cwmh.py:
```diff
@@ -0,0 +1,207 @@
+        super().__init__(target, proposal=proposal, scale=scale,
+                         initial_point=initial_point, **kwargs)
+        # Set scale
+        if isinstance(scale, Number):
+            self.scale = np.ones(self.dim)*scale
+            self._scale_temp = np.ones(self.dim)*scale
+        elif isinstance(scale, np.ndarray):
+            self.scale = scale
+            self._scale_temp = scale.copy()
```
We may want to include this in scale setter method for robustness?

amal-ghamdi replied:
good idea! done

nabriis commented on cuqi/experimental/mcmc/_cwmh.py:
```diff
@@ -0,0 +1,207 @@
+    def validate_target(self):
+        pass # All targets are valid
+
+    @ProposalBasedSamplerNew.proposal.setter 
```
Consider adding a `#TODO` here related to issue 378. This overwriting of the proposal setter is a design we may reconsider.

amal-ghamdi replied:
done

nabriis commented on cuqi/experimental/mcmc/_cwmh.py:
```diff
@@ -0,0 +1,207 @@
+            raise ValueError(fail_msg)
+
+    def step(self):
+        """ Perform one step of the sampler by transitioning the current point
```
I believe we can push this generic docstring to the base class. Other classes will inherit the docstring automatically.

amal-ghamdi replied:
yes, that is great. I removed the doc string in step, tune, set_state and get_state.

nabriis commented on cuqi/experimental/mcmc/_cwmh.py:
```diff
@@ -0,0 +1,207 @@
+        return acc
+
+    def tune(self, skip_len, update_count):
+        """ Tune the parameters of the sampler. This method is called after each
```
Same comment about putting the generic docstring to base class

nabriis commented on tests/experimental/test_mcmc.py:
```diff
@@ -103,6 +114,31 @@ def test_MALA_regression_warmup(target: cuqi.density.Density):
 
+# # ============ CWMH ============
+
+@pytest.mark.parametrize("target", targets)
+def test_CWMH_regression_sample(target: cuqi.density.Density):
+    """Test the CWMH sampler regression."""
+    sampler_old = cuqi.sampler.CWMH(target, scale=np.ones(target.dim))
+    sampler_new = cuqi.experimental.mcmc.CWMHNew(target,
+                                                 scale=np.ones(target.dim))
+    assert_true_if_sampling_is_equivalent(sampler_old, sampler_new,
+                                          Ns=10,
+                                          old_idx=[0, -1],
+                                          new_idx=[1, -1])
+
+@pytest.mark.parametrize("target", targets)
+def test_CWMH_regression_warmup(target: cuqi.density.Density):
+    """Test the CWMH sampler regression."""
+    sampler_old = cuqi.sampler.CWMH(target, scale=np.ones(target.dim))
+    sampler_new = cuqi.experimental.mcmc.CWMHNew(target,
+                                                 scale=np.ones(target.dim))
+    Ns = 100 if target.dim < 50 else 20
+    assert_true_if_warmup_is_equivalent(sampler_old, sampler_new,
+                                        Ns=Ns,
+                                        strategy="MH_like",
+                                        old_idx=[0, -1],
+                                        new_idx=[1, None])
```
Instead of old_idx and new_idx, which implicitly defines a range of values. I suggest for readabiltiy that we define an explicit range that we compare for the two samplers. E.g. one could compare [1..N] with [0..N-1] or similar. I think it would make the test much easier to understand.

amal-ghamdi replied:
I agree, this makes it clearer. I actually thought about this, but it can make it tricky to set default sample size and default range since the range depends on the sample size. Also setting Ns and range separately, adds overhead on the developer to make them consistent, i.e. sat for the same N.

nabriis replied:
Right. I think I got a bit confused. Perhaps this approach of using the indices for each set of samples makes sense. It may be helpful to add docstring with the parameters of this test.

amal-ghamdi replied:
Great, I added documentation for these parameters

nabriis commented on cuqi/experimental/mcmc/_cwmh.py:
```diff
@@ -0,0 +1,207 @@
+        # Propose state x_i_star used to update
+        # each component of x_t step by step
+        x_t = self.current_point.copy()
+        x_star = self.current_point.copy()
+        target_eval_t = self.current_target
+        if isinstance(self.proposal,cuqi.distribution.Distribution):
+            x_i_star = self.proposal(
+                location= self.current_point, scale=self.scale).sample()
+        else:
+            x_i_star = self.proposal(self.current_point, self.scale) 
+        acc = np.zeros(self.dim)
+
+        for j in range(self.dim):
+            # propose state
+            x_star[j] = x_i_star[j]
+
+            # evaluate target
+            target_eval_star = self.target.logd(x_star)
+
+            # ratio and acceptance probability
+            ratio = target_eval_star - target_eval_t  # proposal is symmetric
+            alpha = min(0, ratio)
+
+            # accept/reject
+            u_theta = np.log(np.random.rand())
+            if (u_theta <= alpha):
+                x_t[j] = x_i_star[j]
+                target_eval_t = target_eval_star
+                acc[j] = 1
+            else:
+                pass
+                # x_t[j]       = x_t[j]
+                # target_eval_t = target_eval_t
+            x_star = x_t.copy()
+
+        self.current_target = target_eval_t
+        self.current_point = x_t
+        # NEW: update return
+        # return x_t, target_eval_t, acc
+        return acc
```
In a way I like that you copied the sampler "step" explicitly. But since we are comparing with regression tests, I think it would be worthwhile to "clean up" some of the comments and improve the overall code quality of the step method in this new module.

Same comment holds for the warmup method.

amal-ghamdi replied:
Nice, I updated the code for better readability, style and documentation

nabriis replied:
Thanks! It looks much better now.

nabriis reviewed (changes requested):
Hi @amal-ghamdi. Thanks for adding the CWMH sampler in the module. The suggested changes look good overall, I simply had some minor comments in relation to the testing framework. In addition, I also had a comment on cleaning up the step and tune methods even more compared to the Sampler module.

amal-ghamdi commented:
Thank you @nabriis for your helpful review! I addressed all your comments except the one about the dimension. We can talk about it today. I also added a comment about your suggested test change: from index to range.

nabriis commented on cuqi/experimental/mcmc/_cwmh.py:
```diff
@@ -0,0 +1,222 @@
+        # Propose a sample x_all_components from the proposal distribution
+        # for all the components
+        target_eval_t = self.current_target
+        if isinstance(self.proposal,cuqi.distribution.Distribution):
+            x_all_components = self.proposal(
+                location= self.current_point, scale=self.scale).sample()
+        else:
+            x_all_components = self.proposal(self.current_point, self.scale)
```
This may be out of scope for this PR, but I wonder if we should get rid of the callable proprosal also? And just use distributions. Could create issue for this.

amal-ghamdi replied:
good point, I added it to #373

nabriis reviewed (approved):
Thanks @amal-ghamdi. Looks good to me! I only had some small comments. I can approve it from here.

amal-ghamdi commented:
Thank you @nabriis for your second review! I updated the PR to reflect your suggestions.

jakobsj commented on cuqi/experimental/mcmc/_cwmh.py:
```diff
@@ -0,0 +1,221 @@
+        The signature of the callback function is
+        `callback(sample, sample_index)`, where `sample` is the current sample
+        and `sample_index` is the index of the sample.
+        An example is shown in demos/demo31_callback.py.
```
Is the way callbacks are illustrated in demo31 still relevant for the new samplers? Maybe add an issue to update the demo once the samplers have replaced the old samplers.

amal-ghamdi replied:
good point, I recorded this in #378

jakobsj commented on cuqi/experimental/mcmc/_cwmh.py:
```diff
@@ -0,0 +1,221 @@
+        self.scale = scale
+
+        # set initial acceptance rate
+        self._acc = [np.ones((self.dim))]
```
Is `dim` still available?  I understood from the PR comments it had been removed?

amal-ghamdi replied:
`dim` as parameter to the initializer is removed. But we still have property `dim` extracted from the target dim.

jakobsj commented on cuqi/experimental/mcmc/_cwmh.py:
```diff
@@ -0,0 +1,221 @@
+                f"{cuqi.density.Density.__class__.__name__}")
+
+    @ProposalBasedSamplerNew.proposal.setter
+    # TODO. Check if we can refactor this.
```
Do we have an issue to remember this? If not, add one?

amal-ghamdi replied:
Thank you for pointing this out, I recorded it in #373

jakobsj reviewed (approved):
Thanks @amal-ghamdi , from a quick look through I think it looks great and only have a couple of tiny questions added as individual comments, but approving now.

amal-ghamdi commented:
@jakobsj , Many thanks for your review! I addressed your comments, feel free to resolve the conversations if you think the comments are addressed properly.
